File: UWP/Shiba/Shiba.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Diagnostics;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using Windows.ApplicationModel.Core;
using Windows.UI.Core;
using Shiba.CommonProperty;
using Shiba.Controls;
using Shiba.ExtensionExecutors;
using Shiba.Scripting;
using Shiba.Scripting.Runtime;
using Console = Shiba.Scripting.Runtime.Console;

namespace Shiba
{
    public class ShibaApp
    {
        private ShibaApp(Action<ShibaConfiguration> action)
        {
            action?.Invoke(Configuration);
            var assemblies = AppDomain.CurrentDomain.GetAssemblies().ToList();
            ViewMappings = assemblies
                .Where(item => item.GetCustomAttributes<ExportMapperAttribute>()?.Any() == true)
                .SelectMany(item => item.GetCustomAttributes<ExportMapperAttribute>()).ToList().AsReadOnly();
        }

        public ReadOnlyCollection<ExportMapperAttribute> ViewMappings { get; }
        internal Dictionary<string, View> Components { get; } = new Dictionary<string, View>();
        public ShibaConfiguration Configuration { get; } = new ShibaConfiguration();
        public static ShibaApp Instance { get; protected set; }
        internal View AppComponent { get; set; }

        public void AddConverter(string name, Func<List<object>, object> converter)
        {
            Configuration.NativeConverters.Add(name, converter);
        }

        public static void Init(Action<ShibaConfiguration> action = null)
        {
            Instance = new ShibaApp(c =>
            {
                c.PlatformType = "UWP";
                c.ScriptRuntime.AddObject("shibaStorage", new Storage());
                c.ScriptRuntime.AddObject("console", new Console());
                c.CommonProperties.Add(new GridProperty());
                c.CommonProperties.Add(new RelativeProperty());
                c.ExtensionExecutors.Add(new BindingExecutor());
                c.ExtensionExecutors.Add(new JsonExecutor());
                action?.Invoke(c);
            }); 
        }
    }

    public class ShibaConfiguration
    {
        public IScriptRuntime ScriptRuntime { get; set; } = new DefaultScriptRuntime();
        public string PlatformType { get; set; } = "Windows";

        public Dictionary<string, Func<List<object>, object>> NativeConverters { get; set; } =
            new Dictionary<string, Func<List<object>, object>>();

        public List<IExtensionExecutor> ExtensionExecutors { get; } = new List<IExtensionExecutor>();

        public List<ICommonProperty> CommonProperties { get; } = new List<ICommonProperty>();
    }
}

File: Android/shiba/src/main/java/org/shibajs/shiba/Shiba.kt
package org.shibajs.shiba

import android.content.Context
import android.util.ArrayMap
import android.util.Log
import org.shibajs.shiba.commonProperty.GridProperty
import org.shibajs.shiba.commonProperty.ICommonProperty
import org.shibajs.shiba.extensionExecutor.BindingExecutor
import org.shibajs.shiba.extensionExecutor.IExtensionExecutor
import org.shibajs.shiba.extensionExecutor.JsonExecutor
import org.shibajs.shiba.mapper.*
import org.shibajs.shiba.scripting.DefaultScriptRuntime
import org.shibajs.shiba.scripting.IScriptRuntime
import org.shibajs.shiba.scripting.runtime.Http
import org.shibajs.shiba.scripting.runtime.Storage
import org.shibajs.shiba.type.View
import org.liquidplayer.javascript.JSFunction
import org.liquidplayer.javascript.JSObject

internal typealias NativeView = android.view.View
internal typealias ShibaView = View

class ShibaConfiguration {
    var scriptRuntime: IScriptRuntime = DefaultScriptRuntime()
    var platformType = "Android"
    val commonProperties = ArrayList<ICommonProperty>()
    val extensionExecutors = ArrayList<IExtensionExecutor>()
    val nativeConverter = androidx.collection.ArrayMap<String, ((List<Any?>) -> Any?)>()
}

object Shiba {
    internal var appComponent: ShibaView? = null
    val viewMapping = ArrayMap<String, IViewMapper<*>>()
    val configuration = ShibaConfiguration()
    internal val components = ArrayMap<String, ShibaView>()

    fun init(application: Context) {
        addRenderer("stack", StackMapper())
        addRenderer("text", TextMapper())
        addRenderer("input", InputMapper())
        addRenderer("list", ListMapper())
        addRenderer("grid", GridMapper())
        addRenderer("button", ButtonMapper())
        addExtensionExecutor(BindingExecutor())
        addExtensionExecutor(JsonExecutor())
        configuration.commonProperties.add(GridProperty())
        if (configuration.scriptRuntime is DefaultScriptRuntime) {
            (configuration.scriptRuntime as DefaultScriptRuntime).addObject("shibaStorage") {
                Storage(application, it)
            }
            (configuration.scriptRuntime as DefaultScriptRuntime).addObject("http") {
                Http(it)
            }
            (configuration.scriptRuntime as DefaultScriptRuntime).addObject("console") {
                object : JSObject(it) {
                    init {
                        property("log", object : JSFunction(it, "log") {
                            fun log(msg: String) {
                                Log.i("Javascript", msg)
                            }
                        })
                    }
                }
            }
        }
    }

    fun addConverter(name: String, converter: ((List<Any?>) -> Any?)) {
        configuration.nativeConverter[name] = converter
    }

    fun addRenderer(name: String, mapper: IViewMapper<*>) {
        viewMapping[name] = mapper
    }

    fun addExtensionExecutor(executor: IExtensionExecutor) {
        configuration.extensionExecutors.add(executor)
    }
}